samples: tfm_integration: Extend PSA cryptography demo

urutva opened the pull request:
This PR:
- Renames `psa_level_1` sample to `psa_crypto`
- Extends the psa_crypto sample's use of the PSA Cryptography 1.0 API to demonstrate the following functionality:
  - Generate a persistent key (secp256r1)
  - Display the public key based on the private key above
  - Calculate the SHA256 hash of a payload
  - Sign the hash with the persistent key
  - Verify the signature using the public key
  - Destroy the key
  - Generate 256 bytes of random data
- Enable PSA IPC mode in psa_crypto sample
- Fix docs in `tfm_ipc` sample

Signed-off-by: Kevin Townsend <kevin.townsend@linaro.org>

microbuilder committed:
samples: tfm_integration: Extend PSA cryptography demo

This commit extends the psa_level_1 sample's use of the PSA
Cryptography 1.0 API to demonstrate the following functionality:

- Generate a persistent key (secp256r1)
- Display the public key based on the private key above
- Calculates the SHA256 hash of a payload
- Signs the hash with the persistent key
- Verifies the signature using the public key
- Destroys the key

Signed-off-by: Kevin Townsend <kevin.townsend@linaro.org>

frkv commented on samples/tfm_integration/psa_level_1/Kconfig:
```diff
@@ -19,4 +19,19 @@ config PSA_SHELL
+	help
+	  A static key value will be used for the private key.
+
+config PRIVATE_KEY_RANDOM
```
I would set the random key as the default here, as it is the safest option

urutva replied:
@frkv @microbuilder 
The config `PRIVATE_KEY_RANDOM` is not used in the code. When config `PRIVATE_KEY_STATIC` is enabled, a static key is used otherwise the key data is generated using `psa_generate_random`. By default, `PRIVATE_KEY_STATIC` is disabled, therefore random key data. I'd propose to remove the config option `PRIVATE_KEY_RANDOM`.

urutva replied:
Sorry, I misunderstood how `choice` config option worked. Reading https://docs.zephyrproject.org/latest/guides/kconfig/setting.html#configuring-choices made it clear. I'll add `PRIVATE_KEY_RANDOM` as the default choice.

frkv commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -245,43 +272,64 @@ Sample Output
+      00000200 19 55 64 59 18 95 0E 74 90 36 26 6F E7 49 56 2E .UdY...t.6&o.IV.
+      00000210 E8 0E 8D 4D C4 1C 24 1A 52 D5 69 23 DA BE B7 1F ...M..$.R.i#....
+      00000220 0B                                              .
+      [00:00:05.784,000] <inf> app: Persisting SECP256R1 key as #1
```
Output would maybe be a little more readable if there was some way to inject some "air" between the informational strings and the hex output

urutva replied:
I've added a new line between informational strings and hex output. Hope that helps.

frkv commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -25,6 +25,27 @@ multi-thread application.
+
+Demonstrates the following workflow:
+
+- Generate a permanent persistent key (prime256v1 or ecdsa-with-SHA256)
```
"Permanent persistent key" gives a possible miscommunication with a trait of Key Identifiers in PSA. Key Identifiers are explained as "permanent names for persistent keys". Also used in this scope it seems like a double statement. I think  "persistent key" should suffice

See: https://armmbed.github.io/mbed-crypto/html/overview/functionality.html#key-identifiers

I think this needs to change to the the SECG naming "secp256r1" instead of the ANSI x9.62 "prime256v1"
Both legacy mbed TLS APIs and PSA Crypto APIs use the SECG naming. 

I don't understand what "or ecdsa-with-SHA256" means in this context. There is a possibility that this statement gives a miscommunication about key usage flags that can be set with the key. 

E.g: 
https://armmbed.github.io/mbed-crypto/html/api/keys/policy.html#c.PSA_KEY_USAGE_SIGN_MESSAGE
https://armmbed.github.io/mbed-crypto/html/api/keys/policy.html#c.PSA_KEY_USAGE_VERIFY_MESSAGE

urutva replied:
I've updated both code and docs to reflect SECG naming convention. The wording `ecdsa-with-SHA256` indicate usage of the key as you said. I've updated the description as `Generate a persistent key: secp256r1 (usage: ecdsa-with-SHA256)`. Is this better?

microbuilder commented:
Please also enable `CONFIG_TFM_IPC=y` in the config file to default to IPC instead of Library mode for this sample.

urutva committed:
samples: tfm_integration: Enable PSA IPC mode

Enable PSA IPC mode in psa_level_1 sample.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@linaro.org>

ioannisg commented on samples/tfm_integration/psa_level_1/prj.conf:
```diff
@@ -11,6 +11,7 @@ CONFIG_LOG_DEFAULT_LEVEL=4
 #CONFIG_PSA_SHELL=y
 
 CONFIG_BUILD_WITH_TFM=y
+CONFIG_TFM_IPC=y
```
I am fine with this, but i think we really need to change the sample name, since level_1 refers, i guess to the isolation level? and with IPC we can have any level, basically...

microbuilder replied:
I agree ... the name has bothered me for a while. I'll try to come up with something more appropriate with PSA since it's basically showing how to use the various PSA APIs.

microbuilder replied:
@ioannisg @oyvindronningstad @urutva What do you think of renaming this `psa_crypto` since the main goal of the demo is to show off how to use the PSA Crypto API? The level 1 label really no longer has any meaning to this sample as is.

The documentation would also need to be updated anywhere `psa_lvl_1` is present, of course, but I think that still fits in the intention of this PR to improve this sample.

urutva committed:
samples: tfm_integration: Rename sample psa_level_1 to psa_crypto

The example in folder `samples/tfm_integration/psa_level_1` contains
code that demonstrate usage of PSA crypto APIs in Zephyr rather than the
PSA levels. Therefore renaming the sample as `psa_crypto`.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@linaro.org>

urutva committed:
samples: tfm_integration: docs: Fix incorrect sample name

The README.rst in `samples/tfm_integration/tfm_ipc` was incorrectly
referring to `psa_level_1`. README.rst has been updated to refer to
`tfm_ipc` example.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@linaro.org>

ioannisg commented on samples/tfm_integration/psa_crypto/prj.conf:
```diff
@@ -11,6 +11,7 @@ CONFIG_LOG_DEFAULT_LEVEL=4
 #CONFIG_PSA_SHELL=y
 
 CONFIG_BUILD_WITH_TFM=y
+CONFIG_TFM_IPC=y
```
@urutva @microbuilder  guess if we do that, then the only sample we'll be testing without IPC is the PS-sample. Just FYI :)

ioannisg reviewed (commented):
Looks on to me.
Will add HW_BLOCK, so we can test in on HW besides QEMU, before merging.

oyvindronningstad commented on samples/tfm_integration/psa_crypto/README.rst:
```diff
@@ -1,30 +1,50 @@
+
+Demonstrates the following workflow:
+
+- Generate a persistent key: secp256r1 (usage: ecdsa-with-SHA256)
+- Display the public key based on the private key data above
+- Calculates the SHA256 hash of a payload
+- Signs the hash with the persistent key
+- Verifies the signature using the public key
+- Destroys the key
```
Nit: Inconsistent verb tense (generate vs generates).

urutva replied:
Fixed

oyvindronningstad commented on samples/tfm_integration/psa_crypto/Kconfig:
```diff
@@ -19,4 +19,20 @@ config PSA_SHELL
+config PRIVATE_KEY_STATIC
+	bool "Static"
+	help
+	  A static key value will be used for the private key.
```
Please mention which key is used.

urutva replied:
I've addd elliptic curve details.

microbuilder replied:
I don't know if @oyvindronningstad  meant the key type (though it's nice to have) ... or where the static key is defined, which is this chunk of code, converting the included .pem key file???

```
	/* secp256r1 private key. */
#if CONFIG_PRIVATE_KEY_STATIC
	/* This value is based on the private key in user.pem,
	 * which can be viewed viw the following command:
	 *
	 *   $ openssl ec -in user.pem -text -noout
	 */
	uint8_t priv_key_data[32] = {
		0x14, 0xbc, 0xb9, 0x53, 0xa4, 0xee, 0xed, 0x50,
		0x09, 0x36, 0x92, 0x07, 0x1d, 0xdb, 0x24, 0x2c,
		0xef, 0xf9, 0x57, 0x92, 0x40, 0x4f, 0x49, 0xaa,
		0xd0, 0x7c, 0x5b, 0x3f, 0x26, 0xa7, 0x80, 0x48
	};
#else
	/* Randomly generate the private key. */
	uint8_t priv_key_data[32] = { 0 };

	psa_generate_random(priv_key_data, sizeof(priv_key_data));
#endif
```

microbuilder replied:
Also, I see a typo in the comment, `viw` = `with` (my fault )

ioannisg commented:
@urutva pls address the minor issues raised by @oyvindronningstad so we can get this PR merged.

ioannisg commented:
I rebase here would re-trigger the documentation build :)

urutva commented:
> I rebase here would re-trigger the documentation build :)

@ioannisg I've fixed doc build failure caused by incorrect tag used in `samples/tfm_integration/psa_protected_storage/README.rst`

```/home/runner/work/zephyr/zephyr/doc/_build/src/samples/tfm_integration/psa_protected_storage/README.rst:37:undefined label: tfm_psa_level_1```
